src/params.rs

```diff
@@ -0,0 +1,142 @@
+//! Fixed size parameters list with optional subparameters.
+
+use core::fmt::{self, Debug, Formatter};
+
+pub(crate) const MAX_PARAMS: usize = 32;
+
+#[derive(Default)]
+pub struct Params {
+    /// Number of subparameters for each parameter.
+    ///
+    /// For each entry in the `params` slice, this stores the length of the param as number of
+    /// subparams at the same index as the param in the `params` slice.
+    ///
+    /// At the subparam positions the length will always be `0`.
+    subparams: [u8; MAX_PARAMS],
+
+    /// All parameters and subparameters.
+    params: [i64; MAX_PARAMS],
+
+    /// Number of suparameters in the current parameter.
+    current_subparams: u8,
+
+    /// Total number of parameters and subparameters.
+    len: usize,
+}
+
+impl Params {
+    /// Returns the number of parameters.
+    #[inline]
+    pub fn len(&self) -> usize {
```

Comment:
I do wonder if we should use `#[inline(always)]` to ensure the inlining, however I'm not sure that it's a good idea, so maybe just `#[inline]` is fine.

Reply:
Inlining is a science in and of itself, though generally `#[inline(always)]` is not recommended afaik. This is also why STD has only `#[inline]` on `Vec::len` for example.

So I think if llvm has a really good reason not to inline it, sure. It should be fine.